openmls/src/group/core_group/past_secrets.rs

```diff
@@ -14,47 +15,92 @@ struct EpochTree {
 /// Can store message secrets for up to `max_epochs`. The trees are added with [`self::add()`] and can be queried
 /// with [`Self::get_epoch()`].
 #[derive(Debug, Serialize, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
 pub struct MessageSecretsStore {
+    // Maximum size of the `past_epoch_trees` list.
     max_epochs: usize,
-    epoch_trees: VecDeque<EpochTree>,
+    // Past message secrets.
+    past_epoch_trees: VecDeque<EpochTree>,
+    // The message secrets of the current epoch.
+    message_secrets: MessageSecrets,
 }
 
 impl MessageSecretsStore {
-    /// Create a new store that can hold up to `max_epochs` message secrets.
-    /// If `max_epochs` is 0, no secret trees will be stored.
-    pub fn new(max_epochs: usize) -> Self {
+    /// Create a new store that can hold up to `max_past_epochs` message secrets.
+    /// If `max_past_epochs` is 0, only the current epoch is being stored.
+    pub(crate) fn new_with_secret(max_epochs: usize, message_secrets: MessageSecrets) -> Self {
         Self {
             max_epochs,
-            epoch_trees: VecDeque::new(),
+            past_epoch_trees: VecDeque::new(),
+            message_secrets,
+        }
+    }
+
+    /// Resize the store.
+    pub(crate) fn resize(&mut self, max_past_epochs: usize) {
+        self.max_epochs = max_past_epochs;
+        while self.past_epoch_trees.len() > self.max_epochs {
+            self.past_epoch_trees.pop_front();
         }
     }
 
     /// Add a secret tree for a given epoch `group_epoch`.
+    /// Note that this does not take the epoch into account and pops out the
+    /// oldest element.
     pub fn add(&mut self, group_epoch: GroupEpoch, message_secrets: MessageSecrets) {
         // Don't store the tree if it's not intended
         if self.max_epochs == 0 {
             return;
         }
-        let GroupEpoch(epoch) = group_epoch;
-        let epoch_tree = EpochTree {
-            epoch,
-            message_secrets,
-        };
-        while self.epoch_trees.len() >= self.max_epochs {
-            self.epoch_trees.pop_front();
+        while self.past_epoch_trees.len() >= self.max_epochs {
```

Comment:
I don't think that will really solve the problem. Both, the `while` and the `debug_assert` are conservative to make sure we don't mess up, because keeping more secrets than we want would be bad. But we'd only move the problem. I tried a couple different versions for rotating/adding and none looked better. Unless you have strong feelings I'd just leave this as is.

Reply:
See the comment about the while loop above.